validator/coordinator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/RecentChainDataPerformanceTracker.java

```diff
@@ -0,0 +1,261 @@
+    if (nodeStartEpoch.isEmpty()) {
+      return;
+    }
+
+    UInt64 currentEpoch = compute_epoch_at_slot(slot);
+    if (!compute_start_slot_at_epoch(currentEpoch).equals(slot)) {
+      return;
+    }
+
+    // Output attestation performance information for current epoch - 2 since attestations can be
+    // included in both the epoch they were produced in or in the one following.
+    if (currentEpoch.isGreaterThanOrEqualTo(nodeStartEpoch.get().plus(UInt64.valueOf(2)))) {
+      statusLogger.performance(
+          getAttestationPerformanceForEpoch(currentEpoch, currentEpoch.minus(UInt64.valueOf(2)))
+              .toString());
+    }
+
+    // Output block performance information for the past BLOCK_PERFORMANCE_INTERVAL epochs
+    if (currentEpoch.isGreaterThanOrEqualTo(BLOCK_PERFORMANCE_EVALUATION_INTERVAL)) {
+      if (currentEpoch.mod(BLOCK_PERFORMANCE_EVALUATION_INTERVAL).equals(UInt64.ZERO)) {
+        statusLogger.performance(
+            getBlockPerformanceForEpochs(
+                    currentEpoch.minus(BLOCK_PERFORMANCE_EVALUATION_INTERVAL), currentEpoch)
+                .toString());
+      }
+    }
+
+    UInt64 epoch =
+        currentEpoch
+            .minus(BLOCK_PERFORMANCE_EVALUATION_INTERVAL)
+            .min(currentEpoch.minus(UInt64.valueOf(2)));
+    clearReduntantSavedSentObjects(epoch);
+  }
+
+  private BlockPerformance getBlockPerformanceForEpochs(
+      UInt64 startEpochInclusive, UInt64 endEpochExclusive) {
+    Set<BeaconBlock> blockInEpoch = getBlocksInEpochs(startEpochInclusive, endEpochExclusive);
+    List<SignedBeaconBlock> sentBlocks =
+        sentBlocksByEpoch.subMap(startEpochInclusive, true, endEpochExclusive, false).values()
+            .stream()
+            .flatMap(Collection::stream)
+            .collect(Collectors.toList());
+    long numberOfIncludedBlocks =
+        sentBlocks.stream()
+            .map(SignedBeaconBlock::getMessage)
+            .filter(blockInEpoch::contains)
+            .count();
+
+    return new BlockPerformance((int) numberOfIncludedBlocks, sentBlocks.size());
+  }
+
+  private AttestationPerformance getAttestationPerformanceForEpoch(
+      UInt64 currentEpoch, UInt64 analyzedEpoch) {
+    checkArgument(
+        analyzedEpoch.isLessThanOrEqualTo(currentEpoch.minus(UInt64.valueOf(2))),
+        "Epoch to analyze attestation performance must be at least 2 epochs less than the current epoch");
+    // Attestations can be included in either the epoch they were produced in or in
+    // the following epoch. Thus, the most recent epoch for which we can evaluate attestation
+    // performance is current epoch - 2.
+    UInt64 analysisRangeEndEpoch = analyzedEpoch.plus(UInt64.valueOf(2));
+
+    // Get included attestations for the given epochs in a map from slot to attestations
+    // included in block.
+    Map<UInt64, List<Attestation>> attestations =
+        getAttestationsIncludedInEpochs(analyzedEpoch, analysisRangeEndEpoch);
```

Comment:
That would mean we'd need to deal with re-orgs which could make this pretty complex. The maximum amount of attestations that can be included in an epoch is 4096, so in two epochs 8k items. I don't think we need to be too concerned about the performance impact of this.

Reply:
I was thinking when we go to actually output the performance metrics, we would filter out any irrelevant results by looking up whether the block hash is considered canonical.  This would be a big rework, so if anything we could possibly tackle this in a follow-up.